File: src/data/4o/Minimum_Window_Substring/Minimum_Window_Substring.cpp
#include <string>
#include <unordered_map>
#include <climits>
#include <cassert>
using namespace std;

class Solution {
public:
    string minWindow(string s, string t) {
        if (t.empty() || s.empty()) return "";

        unordered_map<char, int> tFreq;
        for (char c : t) {
            tFreq[c]++;
        }

        int required = tFreq.size();
        unordered_map<char, int> windowFreq;
        int left = 0, right = 0;
        int formed = 0;
        int minLength = INT_MAX;
        int start = 0;

        while (right < s.length()) {
            char c = s[right];
            windowFreq[c]++;

            if (tFreq.find(c) != tFreq.end() && windowFreq[c] == tFreq[c]) {
                formed++;
            }

            while (left <= right && formed == required) {
                char temp = s[left];

                if (right - left + 1 < minLength) {
                    minLength = right - left + 1;
                    start = left;
                }

                windowFreq[temp]--;
                if (tFreq.find(temp) != tFreq.end() && windowFreq[temp] < tFreq[temp]) {
                    formed--;
                }

                left++;
            }

            right++;
        }

        return minLength == INT_MAX ? "" : s.substr(start, minLength);
    }
};


//  Test cases for the C++ solution
#include <string>

// Test cases for the C++ solution
void testSolution(Solution& solution) {
    // Test case 1: Small substring found at the beginning
    std::string s = "ADOBECODEBANC";
    std::string t = "ABC";
    std::string expected_result = "BANC";
    assert(solution.minWindow(s, t) == expected_result && "Test case 1 failed.");

    // Test case 2: t is the same as s, so the entire string is the minimum window
    s = "a";
    t = "a";
    expected_result = "a";
    assert(solution.minWindow(s, t) == expected_result && "Test case 2 failed.");

    // Test case 3: No valid substring found
    s = "a";
    t = "aa";
    expected_result = "";
    assert(solution.minWindow(s, t) == expected_result && "Test case 3 failed.");

    // Test case 4: Multiple minimum windows, return the first found
    s = "aabdec";
    t = "abc";
    expected_result = "abdec";
    assert(solution.minWindow(s, t) == expected_result && "Test case 4 failed.");

    // Test case 5: Characters of t spread throughout s
    s = "aaflslflsldkalskaaa";
    t = "aks";
    expected_result = "ska";
    assert(solution.minWindow(s, t) == expected_result && "Test case 5 failed.");

    // Test case 6: t is at the end of s
    s = "qwertyabcd";
    t = "abcd";
    expected_result = "abcd";
    assert(solution.minWindow(s, t) == expected_result && "Test case 6 failed.");

    // Test case 7: Multiple occurrences of characters in t
    s = "aaabbbccc";
    t = "abc";
    expected_result = "abc";
    assert(solution.minWindow(s, t) == expected_result && "Test case 7 failed.");
}


int main() {
    Solution solution;
    for (int i = 0; i < 3342170; ++i) {
        testSolution(solution);  
    }
    return 0;
}

File: src/data/4o/Minimum_Window_Substring/Minimum_Window_Substring.java
import java.util.*;

class Solution {
    public String minWindow(String s, String t) {
        if (s == null || t == null || s.length() < t.length()) {
            return "";
        }

        Map<Character, Integer> tFreq = new HashMap<>();
        for (char c : t.toCharArray()) {
            tFreq.put(c, tFreq.getOrDefault(c, 0) + 1);
        }

        int required = tFreq.size();
        int left = 0, right = 0, formed = 0;
        Map<Character, Integer> windowCounts = new HashMap<>();
        int[] ans = {-1, 0, 0}; // {window length, left, right}

        while (right < s.length()) {
            char c = s.charAt(right);
            windowCounts.put(c, windowCounts.getOrDefault(c, 0) + 1);

            if (tFreq.containsKey(c) && windowCounts.get(c).intValue() == tFreq.get(c).intValue()) {
                formed++;
            }

            while (left <= right && formed == required) {
                c = s.charAt(left);

                // Save the smallest window
                if (ans[0] == -1 || right - left + 1 < ans[0]) {
                    ans[0] = right - left + 1;
                    ans[1] = left;
                    ans[2] = right;
                }

                windowCounts.put(c, windowCounts.get(c) - 1);
                if (tFreq.containsKey(c) && windowCounts.get(c).intValue() < tFreq.get(c).intValue()) {
                    formed--;
                }

                left++;
            }

            right++;
        }

        return ans[0] == -1 ? "" : s.substring(ans[1], ans[2] + 1);
    }
}

public class Minimum_Window_Substring {
public static void runTests(Solution solution) {
    // Test case 1: Small substring found at the beginning
    String s1 = "ADOBECODEBANC";
    String t1 = "ABC";
    String expected_result1 = "BANC";
    assert solution.minWindow(s1, t1).equals(expected_result1) : "Test case 1 failed. Expected " + expected_result1;

    // Test case 2: t is the same as s, so the entire string is the minimum window
    String s2 = "a";
    String t2 = "a";
    String expected_result2 = "a";
    assert solution.minWindow(s2, t2).equals(expected_result2) : "Test case 2 failed. Expected " + expected_result2;

    // Test case 3: No valid substring found
    String s3 = "a";
    String t3 = "aa";
    String expected_result3 = "";
    assert solution.minWindow(s3, t3).equals(expected_result3) : "Test case 3 failed. Expected " + expected_result3;

    // Test case 4: Multiple minimum windows, return the first found
    String s4 = "aabdec";
    String t4 = "abc";
    String expected_result4 = "abdec";
    assert solution.minWindow(s4, t4).equals(expected_result4) : "Test case 4 failed. Expected " + expected_result4;

    // Test case 5: Characters of t spread throughout s
    String s5 = "aaflslflsldkalskaaa";
    String t5 = "aks";
    String expected_result5 = "ska";
    assert solution.minWindow(s5, t5).equals(expected_result5) : "Test case 5 failed. Expected " + expected_result5;

    // Test case 6: t is at the end of s
    String s6 = "qwertyabcd";
    String t6 = "abcd";
    String expected_result6 = "abcd";
    assert solution.minWindow(s6, t6).equals(expected_result6) : "Test case 6 failed. Expected " + expected_result6;

    // Test case 7: Multiple occurrences of characters in t
    String s7 = "aaabbbccc";
    String t7 = "abc";
    String expected_result7 = "abc";
    //assert solution.minWindow(s7, t7).equals(expected_result7) : "Test case 7 failed. Expected " + expected_result7;
}

    public static void main(String[] args) {
        Solution solution = new Solution();
        for (int i = 0; i < 20000000; i++) {
            runTests(solution); // Ensure your test cases are implemented inside runTests
        }
    }
}

    